Sources/NIOSSL/SubjectAlternativeName.swift

```diff
@@ -0,0 +1,196 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+
+/// Collection of all Subject Alternative Names from a `NIOSSLCertificate`
+public struct _SubjectAlternativeNames {
+    @usableFromInline
+    internal final class Storage: RandomAccessCollection {
+        public typealias Element = _SubjectAlternativeName
+        
+        @inlinable public var startIndex: Int { 0 }
+        @inlinable public var endIndex: Int { stackSize }
+
+        fileprivate let nameStack: OpaquePointer
+        @usableFromInline internal let stackSize: Int
+
+        internal init(nameStack: OpaquePointer) {
+            self.nameStack = nameStack
+            self.stackSize = CNIOBoringSSLShims_sk_GENERAL_NAME_num(nameStack)
+        }
+        
+        public subscript(position: Int) -> Element {
+            guard let name = CNIOBoringSSLShims_sk_GENERAL_NAME_value(nameStack, position) else {
+                fatalError("Unexpected null pointer when unwrapping SAN value")
+            }
+            
+            let contents = UnsafeBufferPointer(
+                start: CNIOBoringSSL_ASN1_STRING_get0_data(name.pointee.d.ia5),
+                count: Int(CNIOBoringSSL_ASN1_STRING_length(name.pointee.d.ia5))
+            )
+            return .init(nameType: .init(name.pointee.type), contents: .init(collection: self, buffer: contents))
+        }
+
+        deinit {
+            CNIOBoringSSL_GENERAL_NAMES_free(self.nameStack)
+        }
+    }
+    
+    
+    @usableFromInline internal var storage: Storage
+    internal init(nameStack: OpaquePointer) {
+        self.storage = .init(nameStack: nameStack)
+    }
+}
+
+extension _SubjectAlternativeNames: RandomAccessCollection {
+    
+    @inlinable public subscript(position: Int) -> _SubjectAlternativeName {
+        storage[position]
+    }
+    
+    @inlinable public var startIndex: Int {
+        storage.startIndex
+    }
+    
+    @inlinable public var endIndex: Int {
+        storage.endIndex
+    }
```

Comment:
Nit: we still prefer explicit `self` here, even though older code doesn't use it. That applies throughout this file.